Make dynamicsymbols real by default

oliverlee opened the pull request:
SymPy treats all variables as complex by default, but variables used in
sympy/physics/mechanics never need to be complex. The fact that they are
complex by default often causes trouble on differentiation of the resulting
equations of motion because some expressions, e.g. abs(), return different
results for real and complex valued arguments. This changed both time and the
functions to have the real assumption set by default.

dynamicsymbols._t is renamed to dynamicsymbols.t
dynamicsymbols.t is real by default
all functions produced by dynamicsymbols are real by default
dynamicsymbols() returns a tuple instead of list to match symbols()
any additional assumptions can be passed in as kwargs to be passed to symbols()

moorepants committed:
Make dynamicsymbols real by default.

SymPy treats all variables as complex by default, but variables used in
sympy/physics/mechanics never need to be complex. The fact that they are
complex by default often cause trouble on differentiation of the resulting
equations of motion because some expressions, e.g. abs(), return different
results for real and complex valued arguments. This changed both time and the
functions to have the real assumption set by default.

- dynamicsymbols._t is real by default
- all functions produced by dynamicsymbols are real by default
- dynamicsymbols() returns a tuple instead of list to match symbols()
- any additional assumptions can be passed in as kwargs to be passed to
  symbols()

oliverlee committed:
Update tests to use real dynamicsymbols

moorepants commented on sympy/physics/vector/functions.py:
```diff
@@ -593,19 +601,35 @@ def dynamicsymbols(names, level=0):
+        return reduce(diff, [t] * level, syms(t))
 
-dynamicsymbols._t = Symbol('t')
+dynamicsymbols.t = Symbol('t', real=True)
```
We should leave `_t` in place as an alias for backwards compat. I have lots of code that accesses `_t` that will break with this.

oliverlee committed:
Make dynamicsymbols time attribute public

Allow users to access the time symbol used in dynamicsymbols.

oliverlee committed:
Add attribute alias to dynamicsymbols

Add '_t' attribute as an alias for 't' of dynamicsymbols to maintain
backwards compatibility.

oliverlee commented:
@moorepants Do you know how to prevent code from being evaluated in the docstring? For example:

```python
"""
convenience function that is analogous to ``symbols()``. It allows the
user to avoid typing::
 		 
>>> t = symbols('t', real=True)
>>> F = symbols('F', real=True, cls=Function)(t)
"""
```

moorepants commented:
If you put `#doctest: +SKIP` after the line it will not run:

`>>> t = symbols('t', real=True) #doctest: +SKIP`

oliverlee committed:
Update docstring examples that take time derivatives

oliverlee commented:
@moorepants There's a problem on how the documentation is generated. This is a result of defining a ```dynamicsymbol``` class to have two identical attributes. Are you familiar with this?

```
Warning, treated as error:
/home/travis/build/sympy/sympy/doc/src/modules/physics/vector/api/functions.rst:8: WARNING: error while formatting arguments for sympy.physics.vector.dynamicsymbols: '__dynamicsymbols_class' object has no attribute '__bases__'
```

moorepants commented:
No, I haven't seen this. Would have to look into it...

moorepants commented on sympy/physics/mechanics/functions.py:
```diff
@@ -410,7 +410,7 @@ def find_dynamicsymbols(expression, exclude=None):
     set([Derivative(x(t), t)])
     """
-    t_set = set([dynamicsymbols._t])
+    t_set = set([dynamicsymbols.t])
```
Should `_t` be left for backwards compatibility?

moorepants commented on sympy/physics/vector/functions.py:
```diff
@@ -568,44 +569,97 @@ def partial_velocity(vel_list, u_list, frame):
+                feature="Attribute sympy.physics.vector.dynamicsymbols._t",
+                useinstead="attribute sympy.physics.vector.dynamicsymbols.t",
+                deprecated_since_version="0.7.7").warn()
+        return self._get_time_symbol()
```
Oh I see here. You've deprecated it. That looks good.

moorepants commented:
Oliver this looks good, seems like you may need to update with master.

oliverlee commented:
This is the failure from travis:
```
Warning, treated as error:
/home/travis/build/sympy/sympy/doc/src/modules/physics/vector/api/functions.rst:8: WARNING: error while formatting arguments for sympy.physics.vector.dynamicsymbols: '__dynamicsymbols_class' object has no attribute '__bases__'
```

ShaunCurrier commented:
This looks valuable. Is it still a good idea?

moorepants commented:
Yes, this is a good idea. We should finish this up.

moorepants commented:
One thing I've thought about recently is that we should simply have a top level time variable. For example:

```python
from sympy.physics.mechanics import time as t
```

should work. This time variable should have the real assumption. This would be a much clearer way to access the time variable that Vector and Dyadic are aware of.

moorepants commented:
@oliverlee Can you check to allow maintainers to push to your branch on this PR?

oliverlee commented:
>  @oliverlee Can you check to allow maintainers to push to your branch on this PR?

How do I do that?

jksuom commented:
There should be a check box for "Allow edits from maintainers." on the right (below "Lock conversation").

oliverlee commented:
> There should be a check box for "Allow edits from maintainers." on the right (below "Lock conversation").

Thanks!

moorepants committed:
Added get_time and set_time for clearer management of the time variable.

moorepants committed:
Switched to .t instead of .t

moorepants committed:
._t to .t

ShaunCurrier commented:
Thanks all.  I am using the output of Kane's method with a solver to do some system parameter estimation and I believe the real assumption on dynamicsymbols will help the solver.

moorepants commented:
This is close to being merge-able. It's got a few bugs. Will try to get it fixed tomorrow at the scipy sprints.

moorepants commented:
As a work around with current Sympy, just do `dynamicsymbols._t = Symbol('t', real=True)` as the first thing in your scripts that use any physics vector or mechanics functions/classes.

moorepants commented:
@isuruf I'm getting symengine failures here: https://travis-ci.org/sympy/sympy/jobs/253231580

What should we do about it? Should symengine be in the accepted failures?

isuruf commented:
I don't know why there are build failures. This change shouldn't have any consequences. @ShikharJ, will you be able to have a look why the symengine tests fail with this?

@moorepants, if you'd like to get this merged in a hurry, please make symengine a optional test and I'll try to fix this afterwards.

moorepants commented:
This failure: https://travis-ci.org/sympy/sympy/jobs/253231582 does not happen when I build with Sphinx 1.6.3. So https://github.com/sympy/sympy/pull/12965 would likely fix this.

ShikharJ commented on sympy/physics/mechanics/tests/test_lagrange.py:
```diff
@@ -64,7 +64,7 @@ def test_simp_pen():
     q, u = dynamicsymbols('q u')
     qd, ud = dynamicsymbols('q u ', 1)
-    l, m, g = symbols('l m g')
+    l, m, g = symbols('l m g', real=True)
```
@isuruf Though I'm not sure why this happens, but removing `real=True` argument fixes the failure for `test_lagrange.py`. This is probably because at the time of using `sympify` over these objects, the resulting test expression isn't simplified. As far as I know, the `Symbol` class in `SymEngine.py` does take extra arguments but nothing is done of them.

ShikharJ commented on sympy/physics/mechanics/tests/test_particle.py:
```diff
@@ -3,7 +3,7 @@
 
 def test_particle():
-    m, m2, v1, v2, v3, r, g, h = symbols('m m2 v1 v2 v3 r g h')
+    m, m2, v1, v2, v3, r, g, h = symbols('m m2 v1 v2 v3 r g h', real=True)
```
Similar reason as above.

moorepants commented:
I'm thinking the symengine tests should be in allowed failures. I personally don't have the bandwidth (or knowledge) to fix symengine related fails anytime we make changes to the mechanics related modules. I'm worried now about how this could affect future code changes.

isuruf commented:
> I'm thinking the symengine tests should be in allowed failures.

Yes, that would be easier for everybody. Right now the tests for symengine are coupled with other optional deps tests. @asmeurer, how should this be done? new tests and make them allowed_failures?

ShikharJ commented on sympy/physics/mechanics/tests/test_lagrange.py:
```diff
@@ -138,7 +138,7 @@ def test_dub_pen():
     u1, u2 = dynamicsymbols('u1 u2')
     u1d, u2d = dynamicsymbols('u1 u2', 1)
-    l, m, g = symbols('l m g')
+    l, m, g = symbols('l m g', real=True)
```
@isuruf Though I'm not sure why this happens, but removing real=True argument fixes the failure for test_lagrange.py. This is probably because at the time of using sympify over these objects, the resulting test expression isn't simplified. As far as I know, the Symbol class in SymEngine.py does take extra arguments but nothing is done of them.

asmeurer commented:
Add them to the Sage build. It is already in allowed_failures, and runs quickly. Please don't add any new builds to the matrix if at all possible.

moorepants committed:
Moved symengine tests to the allowed failures.

moorepants commented:
Got some kind of sagemath binary download bug:

```
WARNING: The following packages cannot be authenticated!

  sagemath-upstream-binary

E: There were unauthenticated packages and -y was used without --allow-unauthenticated

The command "if [[ "${TEST_SAGE}" == "true" ]]; then sudo apt-add-repository -y ppa:aims/sagemath; sudo apt-get update; sudo apt-get install -y sagemath-upstream-binary; fi" failed and exited with 100 during .
```

Also this on the doc build:

```
/home/travis/build/sympy/sympy/doc/src/modules/physics/vector/api/functions.rst:8: WARNING: error while formatting arguments for sympy.physics.vector.dynamicsymbols: '_dynamicsymbols_class' object has no attribute '__mro__'
```

And now I'm getting this failure that I didn't get before:

```
_______ sympy/physics/mechanics/tests/test_lagrange.py:test_rolling_disc _______

  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/physics/mechanics/tests/test_lagrange.py", line 226, in test_rolling_disc

    12*cos(q2(t))*Derivative(q3(t), t))*Derivative(q1(t), t))/(10*r))

AssertionError
```

moorepants committed:
Fixed the lagrange test failure.

asmeurer commented:
> Got some kind of sagemath binary download bug:

I believe that's why the sage tests are in allowed failures.

moorepants commented:
Any idea how to fix that sphinx build error. I'm thinking it is confused because the class we have is just a class that is implemented to behave like a function.

ShikharJ commented:
@moorepants Please change the following lines:
https://github.com/sympy/sympy/blob/master/sympy/physics/mechanics/tests/test_lagrange.py#L4
and 
https://github.com/sympy/sympy/blob/master/sympy/physics/mechanics/tests/test_particle.py#L1
to import `symbols` as follows:
`from sympy.core.backend import symbols`. I've tried it locally and the tests pass.

oliverlee committed:
Fix symbols import for mechanics tests

moorepants commented:
@ShikharJ You are welcome to open your own PR for that. Just because it passed locally doesn't mean it will on Travis.

czgdp1807 commented:
What should we do about it? Should we go for completing this? Which issue was this PR fixing?

moorepants commented:
Yes, this is a PR that needs to be completed and in a way that's backwards compatible if possible.

sidhu1012 commented:
I would like to try to finish this PR.

moorepants commented:
@sidhu1012 One issue with this is the backwards incompatible changes. It could break code in the wild. This one has languished mostly because of this. A better solution may be to improve documentation and make it easier to set this flag. This needs a bit of discussion. Maybe, read through what's here and then we can talk.

moorepants commented:
There are also two issues this tries to solve. 1) any variables that appear in the equations of motion should be real and 2) we have this hidden variable `dynamicsymbols._t` that shouldn't be hidden, because the user often needs it.

sidhu1012 commented:
> There are also two issues this tries to solve. 1) any variables that appear in the equations of motion should be real 

The best way to solve this could be to improve docstrings and specifically define that variables should be done with assumption `real = True` but this won't be strict and would be up to user to make them real.

Other way could be make a separate `symbols` for physics which is real by default, but this won't be an optimal solution.


> 2) we have this hidden variable `dynamicsymbols._t` that shouldn't be hidden, because the user often needs it.

I think it would be better to store `dynamicsymbols._t` in a time variable or class of some type having a proper docstring and adding that time variable to `__init__.py` of `physics/vector` and making it and all dynamicsymbols real by default.

sidhu1012 commented:
I think this would require many changes and would be backwards compatible.

moorepants commented:
Read this too: https://github.com/sympy/sympy/pull/17714

moorepants commented:
We should submit solutions to 1) and 2) separately.

For the real thing. I think just improving the documentation with recommendations and examples of when this is an issue.

I remember encountering the issue with real once when I used `Abs()` in the definition of something that needed to be differentiated (if I remember correctly). If the arguments of Abs() are real you can get a sensible result for these dynamics problems, but if not you get imaginary values.

sidhu1012 commented:
@moorepants Let's discuss `1) any variables that appear in the equations of motion should be real`  first.

This can be solved by just updating the docstring that all `symbols` used in physics should be `real`  and change all instances of symbol used in examples or docstrings to have `real = True` assumption.

Plus I think we should just update docstring like ` Symbols used in physics must be real or at some point result may not make sense` , or `Unreal symbols doesn't make sense in dynamics`, something like this would suffice along with an example like the one you encountered with `Abs()`.

moorepants commented:
The only issue with changing *all* examples in the docstrings is that it isn't necessary for all the current examples and for examples with a fair number of symbols it clutters things up. We should definitely add some warnings and explanations of this potential issue somewhere as a start. You can even recommend things redifining symbols (and dynamicsymbols):

```
symbols = lambda *args, **kwargs: symbols(*args, **kwargs, real=True)
```

I guess we could even have a real_symbols() and real_dynamicssymbols() function so people can use this if needed without having to type `real=True` everywhere.

sidhu1012 commented:
@moorepants My point is introducing new functions or using `real = True` at some places and not using it in another would complicate things.

One simple adjustment to docstring , `Using real = True in symbols is the best/expected way to define symbols, it's not compulsory but it if not used may give unexpected/wrong result. Example : Abs()` . This much would suffice and it would be up to us to update docstrings which we can do eventually or open a new issue for making symbols real with `easy to fix` label in mechanics and related sub modules.

moorepants commented:
I think adding `real=True` to all symbols and dynamicsymbols in the docs adds unnecessary confusion. I think we only need an explanation in one location and maybe an example of when it is needed. Right now, things like this are here: https://docs.sympy.org/latest/modules/physics/mechanics/advanced.html

sidhu1012 commented:
> I think adding `real=True` to all symbols and dynamicsymbols in the docs adds unnecessary confusion. I think we only need an explanation in one location and maybe an example of when it is needed. Right now, things like this are here: https://docs.sympy.org/latest/modules/physics/mechanics/advanced.html

Yes I agree, we won't need to add `real=True` to all symbols, but I think we don't need to add them explicitly to dynamicsymbols as we should make them real by default.

Instead of adding or changing all docstring, I agree with you that we should add an explanation with an example that when `real = True` is needed. But I don't think we should add something extra, function or anything for this issue. Only a proper explanation would be more than enough.

sidhu1012 commented:
> 2) we have this hidden variable `dynamicsymbols._t` that shouldn't be hidden, because the user often needs it.

#17714 seems to be the appropriate solution . #17714 needs a bit of tuning but seems reasonable to me. I did some changes to #17714 and it seems to pass doctests. If you agree , I could make a PR continuing your work.

moorepants commented:
Feel free to move that forward. I think that the only issue I was hung on was trying to remain bacwkards compatible with `dynamicsymbols._t` which is required because that is in alot of code in the wild.

moorepants commented:
See: https://github.com/search?q=dynamicsymbols._t&type=code

sidhu1012 commented:
> @moorepants  Yes I agree, we won't need to add `real=True` to all symbols, but I think we don't need to add `real=True` explicitly to dynamicsymbols as we should make them real by default.
> 
> Instead of adding or changing all docstring, I agree with you that we should add an explanation with an example that when `real = True` is needed. But I don't think we should add something extra, function or anything for this issue. Only a proper explanation would be more than enough.

@moorepants Any opinion on this?

moorepants commented:
At the minimum some nice explanation and example of when it is needed to set real=True (or other things too like inertia's and masses can't be negative) should be added.

sidhu1012 commented:
> At the minimum some nice explanation and example of when it is needed to set real=True (or other things too like inertia's and masses can't be negative) should be added.

I'll do real part in separate PR.

sidhu1012 commented:
@moorepants The only issue that I'm currently facing is that , `dynamicsymbols._t` can't be updated.  Updating `TIME` is the only way to change symbol from `t`.

moorepants commented:
Yeh, that was an issue. It's unlikely that there is much, if any, code where people do change that variable as was supported. I'm fine deprecating that feature for dynamicsymbols._t.  Maybe browse the code on github that uses dynamicsymbols._t and see if there happens to be any evidence of people using that feature.

We may want to think more carefully about having a keyword argument on relevent methods and functions in physics.vector/mechanics. For example: `KanesMethod(..., time_symbol=Symbol('t'))`, `dynamicsymbols(..., time_symbol=Symbol('t')``. I'll think about it a bit.

sidhu1012 commented:
> Yeh, that was an issue. It's unlikely that there is much, if any, code where people do change that variable as was supported. I'm fine deprecating that feature for dynamicsymbols._t. Maybe browse the code on github that uses dynamicsymbols._t and see if there happens to be any evidence of people using that feature.

I'll check that.


> We may want to think more carefully about having a keyword argument on relevent methods and functions in physics.vector/mechanics. For example: `KanesMethod(..., time_symbol=Symbol('t'))`, `dynamicsymbols(..., time_symbol=Symbol('t')``. I'll think about it a bit.

It is a good idea but we need to be very careful in implementing it because it may confuse the user . Example :- if dynamicsymbol is defined with with some other variable for time but someone is using `t` in KanesMethod then the differentiation would give 0. These features would only be useful for advance user.

moorepants commented:
> It is a good idea but we need to be very careful in implementing it because it may confuse the user . Example :- if dynamicsymbol is defined with with some other variable for time but someone is using t in KanesMethod then the differentiation would give 0. These features would only be useful for advance user.

That's a good reason to keep it as a module level constant.

sidhu1012 commented:
> Yeh, that was an issue. It's unlikely that there is much, if any, code where people do change that variable as was supported. I'm fine deprecating that feature for dynamicsymbols._t. Maybe browse the code on github that uses dynamicsymbols._t and see if there happens to be any evidence of people using that feature.
> 

Scrolled more than 30 pages couldn't find a single instance of that feature.

moorepants commented:
> Scrolled more than 30 pages couldn't find a single instance of that feature.

We can deprecate setting `dynamicsymbols._t = something` for sure then. We could even drop it without deprecating given that it is unlikely used. I think I'm fine with that.

The question is then should we allow changing `TIME`. I guess you will be able to by monkeypatching if you know what you are doing, but we don't have to advertise it necessarily.

sidhu1012 commented:
@moorepants I won't change `dynamicsymbols._t` . It would work as it is working but user won't be able to change it.

`TIME` would work like dynamicsymbols._t used to work and would be changeable by using monkey patching.

sidhu1012 commented:
@moorepants About the real part, making dynamicsymbols and `.t` real by default won't be an appropriate solution as a lot of test cases would fails, especially the derivative ones. 

> At the minimum some nice explanation and example of when it is needed to set real=True (or other things too like inertia's and masses can't be negative) should be added.

This is what seems right to me.

moorepants commented:
Yes, I think we just need improved documentation that makes some suggestions on making symbols and functions real as well as assumptions on other quantities like mass, inertia as being nonnegative (or even positive), etc.

sidhu1012 commented:
> Yes, I think we just need improved documentation that makes some suggestions on making symbols and functions real as well as assumptions on other quantities like mass, inertia as being nonnegative (or even positive), etc.

Once PR #21027 is merged , I'll improve documentation.
